extensions/amp-access-poool/0.1/poool-impl.js

```diff
@@ -0,0 +1,213 @@
+
+    /** @private {string} */
+    this.accessUrl_ = ACCESS_CONFIG['authorization'];
+
+    /** @private {string} */
+    this.iframeUrl_ = ACCESS_CONFIG['iframe'];
+
+    /** @const @private {!../../../src/service/timer-impl.Timer} */
+    this.timer_ = Services.timerFor(this.ampdoc.win);
+
+    /** @const @private {!../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = Services.xhrFor(this.ampdoc.win);
+
+    /** @const @private {!JsonObject} For shape see PooolConfigDef */
+    this.pooolConfig_ = this.accessSource_.getAdapterConfig();
+
+    /** @private {string} */
+    this.bundleID_ = this.pooolConfig_['bundleID'] || '';
+
+    /** @protected {string} */
+    this.readerID_ = '';
+
+    /** @private {string} */
+    this.itemID_ = this.pooolConfig_['itemID'] || '';
+
+    /** @const {!Element} */
+    this.iframe_ = document.createElement('iframe');
+
+    this.initializeIframe_();
+
+    this.checkMandatoryParams_();
+  }
+
+  /**
+   * @return {!Promise<!JsonObject>}
+   */
+  authorize() {
+    return this.getPooolAccess_()
+        .then(response => {
+          return {access: response.access};
+        }, err => {
+          if (!err || !err.response) {
+            throw err;
+          }
+          const {response} = err;
+          if (response.status !== 402) {
+            throw err;
+          }
+          this.renderPoool_();
+          return {access: false};
+        });
+  }
+
+  /**
+   * @private
+   */
+  initializeIframe_() {
+    this.iframe_.setAttribute('id', 'poool-iframe');
+    this.iframe_.setAttribute('scrolling', 'no');
+    this.iframe_.setAttribute('frameborder', '0');
+    setStyle(this.iframe_, 'width', '100%');
+
+    if (this.pooolConfig_['forceWidget'] == 'unlock') {
+      setStyles(this.iframe_, {
+        'height': '250px',
+        'position': 'fixed',
+        'bottom': '0',
+      });
+    } else {
+      setStyles(this.iframe_, {
+        'height': '500px',
+        'transform': 'translateY(-70px)',
+      });
+    }
+  }
+
+  /**
+   * @private
+   */
+  checkMandatoryParams_() {
+    user().assert(this.bundleID_, 'BundleID is incorrect or not provided.');
+    user().assert(this.itemID_, 'ItemID is not provided.');
+  }
+
+  /**
+   * @return {!Promise<Object>}
+   * @private
+   */
+  getPooolAccess_() {
+    const url = this.accessUrl_ + '&iid=' + this.itemID_;
+    const urlPromise = this.accessSource_.buildUrl(url, false);
+    return urlPromise.then(url => {
+      return this.accessSource_.getLoginUrl(url);
+    }).then(url => {
+      dev().info(TAG, 'Authorization URL: ', url);
+      return this.timer_.timeoutPromise(
+          AUTHORIZATION_TIMEOUT,
+          this.xhr_.fetchJson(url)).then(res => res.json());
+    });
+  }
+
+  /**
+   * @private
+   */
+  renderPoool_() {
+    const pooolContainer = document.getElementById('poool');
+    const urlPromise = this.accessSource_.buildUrl(this.iframeUrl_
+        + '&bi=' + this.pooolConfig_['bundleID']
```

Comment:
Ditto, please use `addParamToUrl` or `encodeURIComponent` for all of the params below.